bindings/profilers/cpu.cc

```diff
@@ -13,6 +13,10 @@ namespace dd {
 
 static constexpr size_t k_sample_buffer_size = 100;
 
+static void cleanupProfiler(void* data) {
+   static_cast<CpuProfiler *>(data)->StopAndWaitThread();
```

Comment:
I think the _correct_ thing to do might actually be to `delete` the CpuProfiler instance. If the thread the cpu profiler is _for_ is getting torn down, we shouldn't be keeping the cpu profiler for it, right? Actually, looking at it a bit closer, I'm a bit confused how CpuProfiler can extend from Nan::ObjectWrap and _not_ get cleaned up automatically when the environment is torn down. 🤔

Reply:
yeah, I hesitated with deleting the profiler, but since I have no idea of how the lifetime of objects is managed I wanted to avoid the profiler being deleted twice in some particular cases.
Stopping the profiler is safer: it can be called multiple times.

Reply:
Can you try spinning this up in worker_threads repeatedly and see what the behaviour is of deleting versus this stop-and-wait thing? Does it leak memory with stop-and-wait? Does it crash with delete?

Reply:
I looked at this more closely, here are my findings:
* native object destructor is invoked only when JS wrapper object is garbage collected, this can possibly never happen, be it in a main JS thread or worker JS thread. This can easily lead to memory leak in worker threads if native objects has allocated memory (string/vector/...)
* I think calling `delete` in the `AddEnvironmentCleanupHook` callback is ok, because no GC will run after
* On a simple example, it seems that native objects wrapped with napi/node-addon-api are destructed during env shutdown

WDYT about migrating to napi/node-addon-api ? This would also allow to compile only one version of the profiler addon instead of one for each node version.

Reply:
I don't think napi is possible due to it concealing the V8 interface which we need access to in order the activate the profilers and deal with the output objects. I tried converting to it once before but couldn't find a way to make it work. 😞

Reply:
Do you remember what was the issue with napi ?
Isn't it possible to use v8 API with napi ?(you would need to compile for each node version, but is there any other potential issue?)

Reply:
Napi wraps V8 objects and interfaces with no way to unwrap. So, for example, you can't actually get at the isolate and therefore can't get at the profilers either. _Technically_ you can mix napi and direct v8.h access by using `v8::Isolate::GetCurrent()` within a napi function, but the outputs would not be possible to return due to not being in napi format and not having a direct conversion system. You can manually walk and convert everything, but at that point you've already lost the version independence benefit of napi.

Basically, you can _kind of_ make it work, but it's much more work than just using nan and loses basically any benefits using napi would have.

Reply:
Side note: this is part of why I was considering contributing a pprof-encoding profiler to V8 or Node.js, that way we could have a fairly minimal interface exposed through napi rather than needing to handle the whole tree conversion.

Reply:
Thanks for all the context, I better understand the hurdles to migrate to napi